File: src/main/java/com/kishlaly/ta/utils/Bars.java
package com.kishlaly.ta.utils;

import com.kishlaly.ta.model.Quote;
import org.ta4j.core.BarSeries;
import org.ta4j.core.BaseBarSeries;

import java.time.Duration;
import java.time.ZonedDateTime;
import java.util.List;

import static com.kishlaly.ta.utils.Dates.getTimeInExchangeZone;

/**
 * @author Vladimir Kishlaly
 * @since 26.11.2021
 */
public class Bars {

    public static BarSeries build(List<Quote> quotes) {
        BarSeries initialBarSeries = new BaseBarSeries();
        quotes.forEach(quote -> {
            try {
                ZonedDateTime timestamp = getTimeInExchangeZone(quote.getTimestamp(), quote.exchangeTimezome);
                initialBarSeries.addBar(
                        Duration.ofMinutes(getBarDurationInMinutes()),
                        timestamp,
                        quote.getOpen(),
                        quote.getHigh(),
                        quote.getLow(),
                        quote.getClose(),
                        quote.getVolume());
            } catch (Exception e) {
                System.out.println(e.getMessage());
            }
        });
        return initialBarSeries;
    }

    public static int getBarDurationInMinutes() {
        switch (Context.timeframe) {
            case DAY:
                return Context.workingTime;
            case WEEK:
                return Context.workingTime * 5;
            case HOUR:
            default:
                return 60;
        }
    }

}

File: src/main/kotlin/com/kishlaly/ta/utils/Bars.kt
package com.kishlaly.ta.utils

import com.kishlaly.ta.model.AbstractModelJava
import com.kishlaly.ta.model.Quote
import com.kishlaly.ta.model.TimeframeJava
import org.ta4j.core.BarSeries
import org.ta4j.core.BaseBarSeries
import java.time.Duration

class Bars {

    companion object {

        fun build(quotes: List<Quote>): BarSeries {
            val initialBarSeries = BaseBarSeries()
            quotes.forEach {
                try {
                    val timestamp = Dates.getTimeInExchangeZone(it.timestamp, AbstractModelJava.exchangeTimezome)
                    initialBarSeries.addBar(
                            Duration.ofMinutes(BarsJava.getBarDurationInMinutes().toLong()),
                            timestamp,
                            it.open,
                            it.high,
                            it.low,
                            it.close,
                            it.volume)
                } catch (e: Exception) {
                    println(e.message)
                }
            }
            return initialBarSeries
        }

        fun getBarDurationInMinutes(): Int {
            return when (ContextJava.timeframe) {
                TimeframeJava.DAY -> ContextJava.workingTime
                TimeframeJava.WEEK -> ContextJava.workingTime * 5
                TimeframeJava.HOUR -> 60
                else -> 60
            }
        }

    }

}